src/serde_support.rs

```diff
@@ -0,0 +1,330 @@
+                }
+            }
+            Err(err) => {
+                return Err(de::Error::custom(format!(
+                    "Failed to deserialize SignatureDouble: {err:?}"
+                )))
+            }
+        }
+        let sig = SignatureDouble::from_bytes(&bytes).map_err(|err| {
+            de::Error::custom(format!(
+                "Failed to deserialize SignatureDouble: {err:?}"
+            ))
+        })?;
+        Ok(sig)
+    }
+}
+
+impl Serialize for PublicKeyVarGen {
+    fn serialize<S: Serializer>(
+        &self,
+        serializer: S,
+    ) -> Result<S::Ok, S::Error> {
+        let s = bs58::encode(self.to_bytes()).into_string();
+        serializer.serialize_str(&s)
+    }
+}
+
+impl<'de> Deserialize<'de> for PublicKeyVarGen {
+    fn deserialize<D: Deserializer<'de>>(
+        deserializer: D,
+    ) -> Result<Self, D::Error> {
+        let s = String::deserialize(deserializer)?;
+        let mut bytes: [u8; Self::SIZE] = [0; Self::SIZE];
+        match bs58::decode(&s).into(&mut bytes) {
+            Ok(n) => {
+                if n != Self::SIZE {
+                    return Err(de::Error::custom(
+                        "failed to deserialize PublicKeyVarGen: invalid byte length",
+                    ));
+                }
+            }
+            Err(err) => {
+                return Err(de::Error::custom(format!(
+                    "Failed to deserialize PublicKeyVarGen: {err:?}"
+                )))
+            }
+        }
+        let pk = PublicKeyVarGen::from_bytes(&bytes).map_err(|err| {
+            de::Error::custom(format!(
+                "Failed to deserialize PublicKeyVarGen: {err:?}"
+            ))
+        })?;
+        Ok(pk)
+    }
+}
+
+impl Serialize for SecretKeyVarGen {
+    fn serialize<S: Serializer>(
+        &self,
+        serializer: S,
+    ) -> Result<S::Ok, S::Error> {
+        let s = bs58::encode(self.to_bytes()).into_string();
+        serializer.serialize_str(&s)
+    }
+}
+
+impl<'de> Deserialize<'de> for SecretKeyVarGen {
+    fn deserialize<D: Deserializer<'de>>(
+        deserializer: D,
+    ) -> Result<Self, D::Error> {
+        let s = String::deserialize(deserializer)?;
+        let mut bytes: [u8; Self::SIZE] = [0; Self::SIZE];
+        match bs58::decode(&s).into(&mut bytes) {
+            Ok(n) => {
+                if n != Self::SIZE {
+                    return Err(de::Error::custom(
+                        "failed to deserialize SecretKeyVarGen: invalid byte length",
+                    ));
+                }
+            }
+            Err(err) => {
+                return Err(de::Error::custom(format!(
+                    "Failed to deserialize SecretKeyVarGen: {err:?}"
+                )))
+            }
+        }
+        let pk = SecretKeyVarGen::from_bytes(&bytes).map_err(|err| {
+            de::Error::custom(format!(
+                "Failed to deserialize SecretKeyVarGen: {err:?}"
+            ))
+        })?;
+        Ok(pk)
```

Comment:
```suggestion
        let s = String::deserialize(deserializer)?;
        let mut bytes: [u8; Self::SIZE] = [0; Self::SIZE];

        let bytes_written = bs58::decode(&s)
            .into(&mut bytes)
            .map_err(serde::de::Error::custom)?;

        if bytes_written == Self::SIZE {
            Ok(SecretKeyVarGen::from_bytes(&bytes).map_err(|err| {
                serde::de::Error::custom(format!("{:?}", err))
            })?)
        } else {
            Err(de::Error::custom(
                "Failed to deserialize SecretKeyVarGen: invalid byte length",
            ))
        }
```

The `.into()` function on bs58 in this case is not a trait impl of the Into trait from Rust, so it is (at least to me) not immediately clear that the actual return value `n` here is not the inferred type, but the length written to the target array. I would suggest making this explicit by using `bytes_written` or `length_written` for the variable name.

Also, I think we can safely assume that the error messages from dusk_bytes or bs58 are informative enough when they appear, so we can just call map_err with serde custom on those error cases.

I also added an example suggestion above, showing how this can be written more concisely